// @plasmicapp/query/dist/index.js
"use client";
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// src/index.tsx
var src_exports = {};
__export(src_exports, {
  HeadMetadataContext: () => HeadMetadataContext,
  PlasmicPrepassContext: () => PlasmicPrepassContext,
  PlasmicQueryDataProvider: () => PlasmicQueryDataProvider,
  addLoadingStateListener: () => addLoadingStateListener,
  isPlasmicPrepass: () => isPlasmicPrepass,
  useMutablePlasmicQueryData: () => useMutablePlasmicQueryData,
  usePlasmicDataConfig: () => usePlasmicDataConfig,
  usePlasmicQueryData: () => usePlasmicQueryData,
  useSWRConfig: () => import_swr2.useSWRConfig,
  wrapLoadingFetcher: () => wrapLoadingFetcher
});
module.exports = __toCommonJS(src_exports);

// src/query-data.tsx
var import_react = __toESM(require("react"));
var import_swr = __toESM(require("swr"));
var __SWRConfig = void 0;
var mutateKeys = (invalidateKey) => {
  if (__SWRConfig) {
    const { cache, mutate } = __SWRConfig;
    (invalidateKey != null ? [invalidateKey] : Array.from(cache.keys())).forEach((key) => {
      mutate(key);
    });
  }
};
function getPlasmicDefaultSWROptions(opts) {
  return {
    revalidateIfStale: !!(opts == null ? void 0 : opts.isMutable),
    revalidateOnFocus: false,
    revalidateOnReconnect: false
  };
}
function usePlasmicQueryData(key, fetcher) {
  const prepassCtx = import_react.default.useContext(PrepassContext);
  const opts = getPlasmicDefaultSWROptions();
  if (prepassCtx) {
    opts.suspense = true;
  }
  const config = (0, import_swr.useSWRConfig)();
  import_react.default.useEffect(() => {
    __SWRConfig = config;
  }, [config]);
  const wrappedFetcher = import_react.default.useMemo(
    () => wrapLoadingFetcher(fetcher),
    [fetcher]
  );
  const resp = (0, import_swr.default)(key, wrappedFetcher, opts);
  if (resp.data !== void 0) {
    return { data: resp.data };
  } else if (resp.error) {
    return { error: resp.error };
  } else {
    return { isLoading: true };
  }
}
function useMutablePlasmicQueryData(key, fetcher, options) {
  const prepassCtx = import_react.default.useContext(PrepassContext);
  const opts = __spreadValues(__spreadValues({}, getPlasmicDefaultSWROptions({ isMutable: true })), options);
  if (prepassCtx) {
    opts.suspense = true;
  }
  const config = (0, import_swr.useSWRConfig)();
  import_react.default.useEffect(() => {
    __SWRConfig = config;
  }, [config]);
  const [isLoading, setIsLoading] = import_react.default.useState(false);
  const fetcherWrapper = import_react.default.useCallback(
    (...args) => __async(this, null, function* () {
      setIsLoading(true);
      try {
        return yield wrapLoadingFetcher(fetcher)(...args);
      } finally {
        setIsLoading(false);
      }
    }),
    [fetcher]
  );
  const laggyDataRef = import_react.default.useRef();
  const { isValidating, mutate, data, error } = (0, import_swr.default)(
    key,
    fetcherWrapper,
    opts
  );
  import_react.default.useEffect(() => {
    if (data !== void 0) {
      laggyDataRef.current = data;
    }
  }, [data]);
  return import_react.default.useMemo(
    () => __spreadValues(__spreadValues({
      isValidating,
      mutate,
      isLoading: data === void 0 && error === void 0 || isLoading
    }, data !== void 0 ? { data } : error === void 0 && laggyDataRef.current ? (
      // Show previous data if available
      { data: laggyDataRef.current, isLagging: true }
    ) : {}), error !== void 0 ? { error } : {}),
    [isValidating, mutate, data, error, isLoading]
  );
}
function PlasmicQueryDataProvider(props) {
  const { children, suspense, prefetchedCache } = props;
  const prepass = import_react.default.useContext(PrepassContext);
  if (prepass) {
    return /* @__PURE__ */ import_react.default.createElement(import_react.default.Fragment, null, children);
  } else {
    return /* @__PURE__ */ import_react.default.createElement(
      import_swr.SWRConfig,
      {
        value: {
          fallback: prefetchedCache != null ? prefetchedCache : {},
          suspense
        }
      },
      children
    );
  }
}
var PrepassContext = import_react.default.createContext(false);
function PlasmicPrepassContext(props) {
  const { cache, children } = props;
  return /* @__PURE__ */ import_react.default.createElement(PrepassContext.Provider, { value: true }, /* @__PURE__ */ import_react.default.createElement(
    import_swr.SWRConfig,
    {
      value: {
        provider: () => cache,
        suspense: true,
        fallback: {}
      }
    },
    children
  ));
}
var usePlasmicDataConfig = import_swr.useSWRConfig;
var loadingCount = 0;
var listeners = [];
function addLoadingStateListener(listener, opts) {
  listeners.push(listener);
  if (opts == null ? void 0 : opts.immediate) {
    listener(loadingCount > 0);
  }
  return () => {
    listeners.splice(listeners.indexOf(listener), 1);
  };
}
function wrapLoadingFetcher(fetcher) {
  return (...args) => __async(this, null, function* () {
    if (loadingCount === 0) {
      listeners.forEach((listener) => listener(true));
    }
    loadingCount += 1;
    try {
      const res = fetcher(...args);
      return isPromiseLike(res) ? yield res : res;
    } finally {
      loadingCount -= 1;
      if (loadingCount === 0) {
        listeners.forEach((listener) => listener(false));
      }
    }
  });
}
function isPromiseLike(x) {
  return !!x && typeof x === "object" && "then" in x && typeof x.then === "function";
}
function isPlasmicPrepass() {
  var _a, _b, _c;
  return !!((_c = (_b = (_a = import_react.default.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED) == null ? void 0 : _a.ReactCurrentDispatcher) == null ? void 0 : _b.current) == null ? void 0 : _c.isPlasmicPrepass);
}
var HeadMetadataContext = import_react.default.createContext({});

// src/index.tsx
var import_swr2 = require("swr");
if (typeof window !== "undefined") {
  const root = window;
  const maybeExistingMutateAllKeys = root.__SWRMutateAllKeys;
  root.__SWRMutateAllKeys = (invalidateKey) => {
    mutateKeys(invalidateKey);
    if (typeof maybeExistingMutateAllKeys === "function") {
      maybeExistingMutateAllKeys(invalidateKey);
    }
  };
}
